notcat.1
.TH NOTCAT 1
.SH NAME
notcat \- print notifications
.SH SYNOPSIS
.B notcat
[\fB\-h\fR | \fB\-\-help\fR]
.br
.B notcat
[\fBclose\fR \fIID\fR | \fBgetcapabilities\fR | \fBgetserverinfo\fR]
.br
.B notcat send
[\fB-aAchiItu\fR \fIVALUE\fR]... [\fB-p\fR] [\fB--\fR] [\fISUMMARY\fR]
[\fIBODY\fR]
.br
.B notcat
[\fB\-se\fR] [\fB\-t\fR \fITIMEOUT\fR] [\fB\-\-capabilities=\fICAP\fR,\fICAP\fR...] \\
.br
       [\fB\-\-on\-notify=\fICMD\fR] [\fB\-\-on\-close=\fICMD\fR] [\fB\-\-on\-empty=\fICMD\fR] \\
.br
       [\fB\-\-\fR] [\fIFORMAT ARGUMENTS\fR]...
.SH DESCRIPTION
.B notcat
is a D-Bus notification daemon which formats and prints received
notifications to standard output, or runs arbitrary subcommands when
notifications are received and deleted.
.PP
The
.I FORMAT
arguments control how notifications are formatted when they are
printed or executed in subcommands.
They support arbitrary strings.
If
.B notcat
finds one of the following format sequences among them, then it will
replace it in-place with the appropriate value, \fBprintf\fR-style.
.TP
\fB%%\fR
A literal "%"
.TP
\fB%i\fR
Notification ID
.TP
\fB%a\fR
App name
.TP
\fB%s\fR
Summary
.TP
\fB%B\fR
Body text, with newlines replaced with spaces to maintain printing
one line per notification
.TP
\fB%e\fR
Expiration timeout in milliseconds; 0 indicates "no timeout", and -1
indicates "timeout is up to the notification server"
.TP
\fB%u\fR
Urgency; either 'LOW', 'NORMAL', or 'CRITICAL'
.TP
\fB%n\fR
Type of notcat event (i.e., which subcommand is being called); either
"notify" or "close".
.TP
\fB%(h:\fINAME\fB)\fR
Hint value with the given
.I NAME
.PP
If no
.I FORMAT
arguments are provided, then
.B notcat
defaults to '\fB%s\fR'.
.PP
If the
.B \-e
argument is supplied, then instead of the formatted arguments, subcommands are invoked with the following environment variables set to the appropriate values:
.TP
$\fINOTE_ID\fR
Notification ID
.TP
$\fINOTE_APP_NAME\fR
App name
.TP
$\fINOTE_SUMMARY\fR
Summary
.TP
$\fINOTE_BODY\fR
Notification body
.TP
$\fINOTE_URGENCY\fR
Urgency; either 'LOW', 'NORMAL', or 'CRITICAL'
.TP
$\fINOTCAT_EVENT\fR
Type of notcat event (i.e., which subcommand is being called); either
"notify", "close", or "empty".
.TP
$\fINOTE_TIMEOUT\fR
Expiration timeout in milliseconds; 0 indicates "no timeout", and -1
indicates "timeout is up to the notification server"
.PP
Arbitrary hints are not yet supported with the
.B \-e
flag.
.B \-e
is useless when used without explicit subcommands, because
\fBnotcat\fR's default behavior only introspects the
.I FORMAT
arguments.
.SH OPTIONS
.TP
\fB\-h\fR, \fB\-\-help\fR
Display help and exit.
.TP
\fB\-s\fR, \fB\-\-shell\fR
Run subcommands within a subshell, using
.IP
$\fISHELL\fR \-c $\fICMD\fR
.TP
\fB\-e\fR, \fB\-\-env\fR
Pass subcommands information about notifications through the
environment, rather than using formatted arguments.
.TP
\fB\-t\fR, \fB\-\-timeout=\fITIMEOUT\fR
Default timeout for notifications received by \fBnotcat\fR, expressed
in milliseconds.
This value must be expressible as a 32-bit integer.
.TP
\fB\-\-capabilities=\fICAP\fR,\fICAP\fR...
Additional capabilities for
.B notcat
to advertise.
By default,
.B notcat
tries to advertise a reasonable set of capabilities, either (if
.B \-e
is not set) based on the
.I FORMAT
arguments, or based on the set of environment variables that
.B notcat
passes to subcommands.
.TP
\fB\-\-on\-notify=\fICMD\fR
Run the command
.I CMD
each time a notification is created or replaced.
By default, the command is invoked via \fBexecve\fR, with its
arguments set to the interpolated
.I FORMAT
arguments provided to \fBnotcat\fR.
.IP
If not provided,
.B notcat
defaults to its built-in '\fBecho\fR'.
.TP
\fB\-\-on\-close=\fICMD\fR
Run the command
.I CMD
each time a notification is closed.
By default, the command is invoked via \fBexecve\fR, with its
arguments set to the interpolated
.I FORMAT
arguments provided to \fBnotcat\fR.
.IP
If not provided, then \fBnotcat\fR's default behavior is to do
nothing.
.TP
\fB\-\-on\-empty=\fICMD\fR
Run the command
.I CMD
whenever the last notification is closed, and no more notifications
remain.
The command is invoked via \fBexecve\fR.
Unlike the \fB\-\-on\-notify\fR and \fB\-\-on\-close\fR, since
the \fB\-\-on\-empty\fR event isn't associated with any particular
notification, it is run without arguments.
.IP
If not provided, then \fBnotcat\fR's behavior is to do nothing.
.TP
\fB\-\-\fR
Stop option parsing.
This may be used in case there are
.I FORMAT
arguments which would otherwise be parsed as options to \fBnotcat\fR.
.SH CLIENT COMMANDS
The
.B notcat
command can be used to interact as a client with an already-running
notification server.
Existing commands are:
.TP
\fBclose\fR \fIID\fR
Close the notification with the given \fIID\fR.
.TP
\fBgetcapabilities\fR
Get the capabilities of the notification server.
.TP
\fBgetserverinfo\fR
Get basic information about the notification server.
.TP
\fBsend\fR
Send a notification to the server.
In addition to any options, \fBsend\fR requires at most two arguments
specifying the notification summary and body, respectively.
If only one argument is given, it is used as the summary.
.PP
There are several options for the
.B send
command, described as follows.
.SS SEND OPTIONS
.PP
.B notcat send
supports a relatively standard set of notification options, with some
allowances for extra flexibility.
These options correspond with behavior found in the D-Bus
Notifications API documentation.
.TP
\fB-a\fR, \fB--app-name=\fINAME\fR
App name.
.TP
\fB-A\fR, \fB--actions=\fIACTION\fB,\fIKEY\fB:\fINAME\fR...
Actions.
Multiple comma-separated actions may be supplied in one arg, and if
an action is provided as a \fIKEY\fR:\fINAME\fR, then \fINAME\fR will
be used in the notification display.
.TP
\fB-c\fR, \fB--category=\fICATEGORY\fR
Notification category, which should be one of the values given in the
D-Bus Notifications API documentation.
.TP
\fB-h\fR, \fB--hint=\fITYPE\fB:\fINAME\fB:\fIVALUE\fR
A hint, which is an arbitrary key-value pair.
.B notcat
uses GLib's type format strings and parser to get values from this
argument.
Most likely types to be useful are: \fBb\fR, boolean; \fBy\fR, byte;
\fBs\fR, string; \fBi\fR, 32-bit int; and \fBu\fR, 32-bit unsigned
int.
.TP
\fB-i\fR, \fB--id=\fIID\fR
Notification ID to replace (if currently in use).
Must be a non-negative integer.
.TP
\fB-I\fR, \fB--icon=\fIICON\fR
Name or path of the icon to display with this notification.
.TP
\fB-p\fR, \fB--print-id\fR
If set,
.B notcat
will print the ID of the notification after sending it.
.TP
\fB-u\fR, \fB--urgency=\fIURGENCY\fR
Urgency of the notification.
May be one of \fBlow\fR, \fBnormal\fR, or \fBcritical\fR.
.SH EXAMPLES
Simple invocation to print notification summaries and bodies as they
arrive:
.IP
\fB$\fR notcat %s %B
.PP
Invocation that has the same behavior as above, but by invoking
.B echo
in a subshell on each notification:
.IP
\fB$\fR notcat \-s '\-\-on-notify=echo $*' %s %B
.PP
Note the trailing '$*' in the command; this is required for the
invoked
.B echo
to receive the args from its calling shell.
.PP
Invocation that has (roughly) the same behavior as above, but using
environment variables this time:
.IP
\fB$\fR notcat \-se '\-\-on-notify=echo $NOTE_SUMMARY $NOTE_BODY'
.PP
This invocation has the disadvantage of being somewhat more verbose,
and also loses the automatic formatting that
.B notcat
applies to the notification body.
However, using \-e can make complex shell scripts run as subcommands
significantly clearer.
.SH AUTHOR
.B notcat
is written by Jack Conger (jpco).
.PP
Both
.B notcat
and this manual page are released under the GNU General Public
License, version 3+.
.SH ERRATA
Actions, markup, categories, and links are not yet supported.
Moreover,
.B notcat
lies and claims that actions are supported.
.PP
Some capabilities will likely never be supported.
In particular, the \fBbody-images\fR, \fBicon-multi\fR,
\fBicon-static\fR, and \fBsound\fR capabilities are outside the
intended design of \fBnotcat\fR.
.PP
Only arbitrary hints of type string, int, boolean, or byte are
understood.
More complex types of hints can only be supported specially.
.SH SEE ALSO
\fBnotify\-send\fR\|(1)
.SH STANDARDS
.B notcat
conforms to version 1.2 of the Desktop Notifications Specification.


notcat.c
/* Copyright 2018 Jack Conger */

/*
 * This file is part of notcat.
 *
 * notcat is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * notcat is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with notcat.  If not, see <http://www.gnu.org/licenses/>.
 *
 * GDBus logic has been liberally adapted from
 * https://github.com/dunst-project/dunst/blob/master/src/dbus.c
 *
 * Desktop notifications API details can be found at
 * https://developer.gnome.org/notification-spec/
 */

#include <glib.h>
#include <gio/gio.h>
#include <stdio.h>

#include "notcat.h"

/*
 * dbussy data type pseudo-boilerplate.  The introspection_xml parsing might
 * actually be the least unreadable way to express this information,
 * horrifyingly.
 */

static GDBusNodeInfo *introspection_data = NULL;

static const char *introspection_xml =
    "<?xml version=\"1.0\" encoding=\"UTF-8\"?>"
    "<node name=\""FDN_PATH"\">"
    "    <interface name=\""FDN_IFAC"\">"

    "        <method name=\"GetCapabilities\">"
    "            <arg direction=\"out\" name=\"capabilities\"    type=\"as\"/>"
    "        </method>"

    "        <method name=\"Notify\">"
    "            <arg direction=\"in\"  name=\"app_name\"        type=\"s\"/>"
    "            <arg direction=\"in\"  name=\"replaces_id\"     type=\"u\"/>"
    "            <arg direction=\"in\"  name=\"app_icon\"        type=\"s\"/>"
    "            <arg direction=\"in\"  name=\"summary\"         type=\"s\"/>"
    "            <arg direction=\"in\"  name=\"body\"            type=\"s\"/>"
    "            <arg direction=\"in\"  name=\"actions\"         type=\"as\"/>"
    "            <arg direction=\"in\"  name=\"hints\"           type=\"a{sv}\"/>"
    "            <arg direction=\"in\"  name=\"expire_timeout\"  type=\"i\"/>"
    "            <arg direction=\"out\" name=\"id\"              type=\"u\"/>"
    "        </method>"

    "        <method name=\"CloseNotification\">"
    "            <arg direction=\"in\"  name=\"id\"              type=\"u\"/>"
    "        </method>"

    "        <method name=\"GetServerInformation\">"
    "            <arg direction=\"out\" name=\"name\"            type=\"s\"/>"
    "            <arg direction=\"out\" name=\"vendor\"          type=\"s\"/>"
    "            <arg direction=\"out\" name=\"version\"         type=\"s\"/>"
    "            <arg direction=\"out\" name=\"spec_version\"    type=\"s\"/>"
    "        </method>"

    "        <signal name=\"NotificationClosed\">"
    "            <arg name=\"id\"         type=\"u\"/>"
    "            <arg name=\"reason\"     type=\"u\"/>"
    "        </signal>"

    "        <signal name=\"ActionInvoked\">"
    "            <arg name=\"id\"         type=\"u\"/>"
    "            <arg name=\"action_key\" type=\"s\"/>"
    "        </signal>"
    "   </interface>"
"</node>";

/*
 * Per-method-call logic.  Obviously, the most interesting is notify().
 */

static void get_capabilities(GDBusConnection *conn, const gchar *sender,
			     const GVariant *params,
			     GDBusMethodInvocation *invocation) {
    GVariantBuilder *builder;
    GVariant *value;

    builder = g_variant_builder_new(G_VARIANT_TYPE("as"));
#if ACTIONS
    g_variant_builder_add(builder, "s", "actions");
#endif
    g_variant_builder_add(builder, "s", "body");
    /*
     * g_variant_builder_add(builder, "s", "body-hyperlinks");
     * g_variant_builder_add(builder, "s", "body-markup");
     */

    value = g_variant_new("(as)", builder);
    g_variant_builder_unref(builder);
    g_dbus_method_invocation_return_value(invocation, value);

    g_dbus_connection_flush(conn, NULL, NULL, NULL);
}

static guint id = 0;

static void notify(GDBusConnection *conn, const gchar *sender,
		   GVariant *params,
		   GDBusMethodInvocation *invocation) {
    /* TODO: un-glibify these allocations; we don't want to rely on glib for
     * all of this */
    gchar *appname = NULL;
    guint replaces_id = 0;
    gchar *summary = NULL;
    gchar *body = NULL;
#if ACTIONS
    Actions *actions = g_malloc0(sizeof(Actions));
#endif
    gint timeout = -1;
    gint urgency = 1;

    {
	GVariantIter *iter = g_variant_iter_new(params);
        GVariant *content;
        GVariant *dict_value;
        int idx = 0;
	while ((content = g_variant_iter_next_value(iter))) {
	    switch (idx) {
	    case 0:
		if (g_variant_is_of_type(content, G_VARIANT_TYPE_STRING))
		    appname = g_variant_dup_string(content, NULL);
		break;
            case 1:
                if (g_variant_is_of_type(content, G_VARIANT_TYPE_UINT32))
                    replaces_id = g_variant_get_uint32(content);
                break;
            case 2: break;  /* icon -- not supported */
            case 3:
                if (g_variant_is_of_type(content, G_VARIANT_TYPE_STRING))
                    summary = g_variant_dup_string(content, NULL);
                break;
	    case 4:
                if (g_variant_is_of_type(content, G_VARIANT_TYPE_STRING))
                    body = g_variant_dup_string(content, NULL);
                break;
            case 5:
#if ACTIONS
                if (g_variant_is_of_type(content, G_VARIANT_TYPE_STRING_ARRAY))
		    actions->actions = g_variant_dup_strv(content, &(actions->count));
#endif
                break;
            case 6:
                if (g_variant_is_of_type(content, G_VARIANT_TYPE_DICTIONARY)) {
		    dict_value = g_variant_lookup_value(content, "urgency",
							G_VARIANT_TYPE_BYTE);
		    if (dict_value) {
			urgency = g_variant_get_byte(dict_value);
                        g_variant_unref(dict_value);
                    }
		    // TODO: custom formats
		    // TODO: category (?)
		    // TODO: support tying arbitrary hints to notifications
		}
		break;
	    case 7:
		if (g_variant_is_of_type(content, G_VARIANT_TYPE_INT32))
		    timeout = g_variant_get_int32(content);
		break;
	    }
	    g_variant_unref(content);
	    idx++;
	}

	g_variant_iter_free(iter);
    }

    guint n_id = replaces_id ? replaces_id : ++id;

#if ACTIONS
    if (actions->count < 1) {
	free_actions(actions);
	actions = NULL;
    }
#endif

    /* At some point this will be more sophisticated, lol */
    Note *note = new_note(n_id, appname, summary, body,
#if ACTIONS
			  actions,
#endif
			  timeout, urgency);
    print_note(note);
    free_note(note);

    GVariant *reply = g_variant_new("(u)", n_id);
    g_dbus_method_invocation_return_value(invocation, reply);
    g_dbus_connection_flush(conn, NULL, NULL, NULL);
}

static void close_notification(GDBusConnection *conn, const gchar *sender,
			       GVariant *params,
			       GDBusMethodInvocation *invocation) {
    guint32 id;
    g_variant_get(params, "(u)", &id);

    // TODO: Do the thing that actually "closes" the notification, aha

    g_dbus_method_invocation_return_value(invocation, NULL);
    g_dbus_connection_flush(conn, NULL, NULL, NULL);
}

static void get_server_information(GDBusConnection *conn, const gchar *sender,
				   const GVariant *params,
			           GDBusMethodInvocation *invocation) {
    GVariant *value;
    value = g_variant_new("(ssss)", "notcat", "jpco", VERSION, "1.2");
    g_dbus_method_invocation_return_value(invocation, value);
    g_dbus_connection_flush(conn, NULL, NULL, NULL);
}

/*
 * Signals.  We may send these for Reasons.
 */

/* void signal_notification_closed(GDBusConnection *conn,
				const guint not_id, const guint reason) {
    GVariant *body;
    GError *err;

    body = g_variant_new("(uu)", not_id, reason);
    err = NULL;

    g_dbus_connection_emit_signal(conn, dbus_client,
				  FDN_PATH, FDN_IFAC, "NotificationClosed",
				  body, &err);
    if (err) {
	g_printerr("Could not close notification: %s\n", err->message);
	g_error_free(err);
    }
} */

/*
 * Scaffolding functions.  These put things together properly.
 */

void handle_method_call(GDBusConnection *conn,
			const gchar *sender,
			const gchar *object_path,
			const gchar *interface_name,
			const gchar *method_name,
			GVariant *params,
			GDBusMethodInvocation *invocation,
			gpointer user_data) {
    if (g_strcmp0(method_name, "GetCapabilities") == 0) {
	get_capabilities(conn, sender, params, invocation);
    } else if (g_strcmp0(method_name, "Notify") == 0) {
	notify(conn, sender, params, invocation);
    } else if (g_strcmp0(method_name, "CloseNotification") == 0) {
	close_notification(conn, sender, params, invocation);
    } else if (g_strcmp0(method_name, "GetServerInformation") == 0) {
	get_server_information(conn, sender, params, invocation);
    } else {
	g_printerr("Unknown method name '%s' received from '%s'\n",
		   method_name, sender);
    }
}

static const GDBusInterfaceVTable interface_vtable = {
    handle_method_call
};

static void on_bus_acquired(GDBusConnection *conn, const gchar *name,
			    gpointer user_data) {
    guint reg_id;
    GError *err = NULL;

    reg_id = g_dbus_connection_register_object(conn, FDN_PATH,
					       introspection_data->interfaces[0],
					       &interface_vtable,
					       NULL, NULL, &err);

    if (reg_id == 0) {
	g_printerr("Urgh, failed to register object: %s\n", err->message);
    }
}

static void on_name_acquired(GDBusConnection *conn, const gchar *name,
			     gpointer user_data) {
    g_printerr("Got name %s on the session bus\n", name);
}

static void on_name_lost(GDBusConnection *conn, const gchar *name,
			 gpointer user_data) {
    g_printerr("Lost name %s on the session bus\n", name);
}

int main(int argc, char **argv) {
    GMainLoop *loop;
    guint owner_id;

    introspection_data = g_dbus_node_info_new_for_xml(introspection_xml,
						      NULL);

    owner_id = g_bus_own_name(G_BUS_TYPE_SESSION,
	    FDN_NAME,
	    G_BUS_NAME_OWNER_FLAGS_NONE,
	    on_bus_acquired,
	    on_name_acquired,
	    on_name_lost,
	    NULL,
	    NULL);

    loop = g_main_loop_new(NULL, FALSE);
    g_main_loop_run(loop);

    g_bus_unown_name(owner_id);

    return 0;
}
